src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -1782,8 +1782,15 @@ protected void eagerDecode() {
 
     @Override
     public boolean equals(final Object o) {
-        if (this == o) return true;
-        if (!(o instanceof SAMRecord)) return false;
+        if (this == o) {
+            return true;
+        }
+        if(o == null) {
+            return false;
+        }
+        if (this.getClass() != o.getClass()) {
```

Comment:
This change needs some real discussion.  @nh13 @yfarjoun @droazen @vadimzalunin - what are your thoughts on this?  With this change a `SAMRecord`, `BAMRecord` and `CRAMRecord` that are all logically equal, will no longer be equal.  I think this is a bad idea - it's more important that `SAMRecord` implement logical equality than to slavishly follow java convention that `equals` require type equality also.